### PR TITLE
165 issue untyped events and event listeners

### LINKED ISSUE #165: 🤔 [ISSUE] - Untyped Events and Event Listners
### Description

The current implementation of the event emitter API lacks strong typing

### Reproduction steps

```bash
No response
```

### Logs

_No response_

### Browsers

_No response_

### OS

_No response_

### PR DESCRIPTION
**Pull request by yassinedorbozgithub**
# Motivation
 The main motivation is enhancing events typing 

Fixes #165

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (Storybook)
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api/src/utils/generics/base-repository.ts [hunk 1]
```diff
@@ -96,11 +104,21 @@ export abstract class BaseRepository<
     hooks?.deleteOne.pre.execute(async function () {
       const query = this as Query<DeleteResult, D, unknown, T, 'deleteOne'>;
       const criteria = query.getQuery();
+      repository.emitter.emit(
+        repository.getEventName(EHook.preDelete),
+        query,
+        criteria,
+      );
       await repository.preDelete(query, criteria);
```
Comment (marrouchi) [inline api/src/utils/generics/base-repository.ts:112]: Let's invert the order (for all)

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: api/src/utils/generics/base-repository.ts (hunk 1 above)
Comment (yassinedorbozgithub) [reply api/src/utils/generics/base-repository.ts:112]: Sure, i will do that :technologist:

**Review by marrouchi — APPROVED** [review]
@yassinedorbozgithub Looks perfect to me 👏 

Let's do PRs in the current extensions we have so that it doesn't break (primarely messenger for now)

**Comment by yassinedorbozgithub** [comment]
> @yassinedorbozgithub Looks perfect to me 👏
> 
> Let's do PRs in the current extensions we have so that it doesn't break (primarely messenger for now)

Sure happy to do that :technologist:

**Review by IkbelTalebHssan — APPROVED** [review]

**Pull request merged**